### PR TITLE
[ExecutionEngine] Remove the logic that updates variables from the EE.

### PR DESCRIPTION
**Pull request by nadavrot**
*Description*:

This PR refactors the execution engine and removes all of the logic that is responsible for updating variables into external helper functions. It does not make sense for the execution engine, that managed the lifetime of the compiled function to also update variables, and once we migrate to placeholders we won't be able to rely on the execution engine to update the variables/placeholders for us, because they won't be tied to the run() method. 

This change removes the runBatch() method and keeps a simple run() method. All of the variable changes happen before the execution happens. This is the API change for the unit tests:

Before:

    EE.run({var1, var2}, {inputs1, inputs2});	 

After:

    updateVariables({var1, var2}, {inputs1, inputs2});
    EE.run();

*Testing*: I ran "ninja check", the mnist training test and the run.sh inference test.
*Documentation*: Updated doxygen.

### CODE REVIEW

File: include/glow/ExecutionEngine/ExecutionEngine.h [hunk 1]
```diff
@@ -78,34 +78,34 @@ class ExecutionEngine final {
...
-  void loadValueFromTensor(Variable *v, Tensor *input);
+  /// Runs a single execution of the function.
+  void run();
 };
 
+//===----------------------------------------------------------------------===//
+//         Helper methods for running the execution engine.
+//===----------------------------------------------------------------------===//
+
+/// This method updates the variables in \p nodes with the tensor content
```
Comment (rdzhabarov) [inline include/glow/ExecutionEngine/ExecutionEngine.h]: nit: \p vars with the tensor ...

File: lib/ExecutionEngine/ExecutionEngine.cpp [hunk 2]
```diff
@@ -55,77 +55,74 @@ void ExecutionEngine::setBackend(Backend *backend) {
...
+    // Extract the n'th slice, that must be a tensor.
+    size_t slc = sampleIdx % dim[0];
+    t.copyConsecutiveSlices(inputs[i], slc);
+  }
+}
+
+void glow::runBatch(ExecutionEngine &EE, size_t iterations,
+                    llvm::ArrayRef<Variable *> vars,
+                    llvm::ArrayRef<Tensor *> inputs) {
+  static size_t trainCounter = 0;
```
Comment (rdzhabarov) [inline lib/ExecutionEngine/ExecutionEngine.cpp]: why `trainCounter` needs to be static? Can you please explain.

File: include/glow/ExecutionEngine/ExecutionEngine.h [hunk 3]
```diff
@@ -78,34 +78,34 @@ class ExecutionEngine final {
...
+/// dimension must be identical.
+void updateVariablesFromBatch(llvm::ArrayRef<Variable *> vars,
+                              llvm::ArrayRef<Tensor *> inputs,
+                              size_t sampleIdx);
+
+/// Runs \p iterations iterations of the function. The method updates a local
+/// counter and future invocations of this method continue running iterations
+/// of the batch at the next available slice.
+/// The method updates the variables in \p vars with the tensors \p inputs.
+void runBatch(ExecutionEngine &EE, size_t iterations,
```
Comment (qcolombet) [inline include/glow/ExecutionEngine/ExecutionEngine.h]: Is the part with the local counter still true?
Also, while you are here, the comment would benefit from more love :).
After reading it over and over, I am not sure how the iterations are applied to the batch size.

File: lib/ExecutionEngine/ExecutionEngine.cpp (hunk 2 above)
Comment (nadavrot) [reply lib/ExecutionEngine/ExecutionEngine.cpp]: I added a big comment that explains why this is a really bad idea to have this global variable. I think that I have a quick fix to get rid of the global variable. Basically, now that this is only a helper function we can change the call sites and make it a parameter.

File: include/glow/ExecutionEngine/ExecutionEngine.h (hunk 3 above)
Comment (nadavrot) [reply include/glow/ExecutionEngine/ExecutionEngine.h]: I updated the comment and got rid of the global variable. 

/// Runs \p iterations iterations of the compiled function. The method updates a
/// global counter and future invocations of this method continue running
/// iterations of the batch at the next available slice.
///
/// The method updates the variables in \p vars with the tensors \p inputs. The
/// shape of the slice has to be identical to the shape of slices in the batch.
/// All dimensions, except for the first (batch) dimension must be identical.
///
/// The variable \p sampleCounter is consumed and updated by the function. This
/// variable records the number of samples that were consumed by the network in
/// previous iterations. The next input to be loaded is
/// (sampleCounter % batchsize).

**Pull request merged**